### ISSUE TITLE
Inefficient scheduling of large jobs on Dask cluster

### ISSUE THREAD

**Issue by Chroxvi** [joblib/joblib#1021]
I am trying to fit a scikit-learn RandomForestClassifier for a multi class classification problem using the Dask backend for joblib. For small problem sizes, i.e. few training samples, the fitting of the individual trees in the model is nicely parallelized. However, for larger problem sizes, the fit of the individual trees are all scheduled on the same worker which is very inefficient due to the lack of parallelization.

A reproducible example to illustrate the problem is:
```python
import dask
from dask.distributed import Client, LocalCluster
import joblib
import sklearn
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification

print(f'dask version: {dask.__version__}')
print(f'joblib version: {joblib.__version__}')
print(f'scikit-learn version: {sklearn.__version__}')

cluster = LocalCluster()
client = Client(cluster)

@dask.delayed
def dask_joblib_fit(model, X, y):
    with joblib.parallel_backend('dask'):
        model.fit(X, y)
        
    return model

X, y = dask.delayed(make_classification, nout=2)(
    n_samples=int(1e6), n_informative=10, n_features=150, n_classes=450, random_state=8811)
model = RandomForestClassifier(n_estimators=20)
model = client.compute(dask_joblib_fit(model, X, y))
```
```python
dask version: 2.11.0
joblib version: 0.14.1
scikit-learn version: 0.22.1
````

If I wait for the `make_classification` task to finish and call `client.processing()`, I get something like:
```python
{'tcp://127.0.0.1:34895': ('dask_joblib_fit-23b412d6-a795-4b22-b425-5471684a0b51',
  '_parallel_build_trees-batch-2502e7138c8e4e1f8936d0c1ce182009',
  '_parallel_build_trees-batch-f85799a9f0954d959aee3b92dcfa976f',
  '_parallel_build_trees-batch-fa9d1d4c562b4093a010d5b41cde4482',
  '_parallel_build_trees-batch-24d0d70e7c1d4ea0be405508a2456575',
  '_parallel_build_trees-batch-81612b45a75d4e52b442cdc7a4f6a464',
  '_parallel_build_trees-batch-5a9ed7b800be41fc860fdac2354a9d0c',
  '_parallel_build_trees-batch-a9bf8a47287e4969817300ce03131ed3',
  '_parallel_build_trees-batch-d7f549c10e054cefb451e5450b8a3139'),
 'tcp://127.0.0.1:39957': (),
 'tcp://127.0.0.1:41333': (),
 'tcp://127.0.0.1:42343': ()}
```
All `_parallel_build_trees-batches` have been schedule on a single worker. However, if I train on 1e4 samples instead of 1e6, I get the something like:
```python
{'tcp://127.0.0.1:35135': ('dask_joblib_fit-783b0f51-eebc-4df6-9d33-dcf644ed7fc5',
  '_parallel_build_trees-batch-e4cc63ea849e4debbccf79f83eeee3b0',
  '_parallel_build_trees-batch-a851f1e8653b462e82ca263246c743c5'),
 'tcp://127.0.0.1:35335': ('_parallel_build_trees-batch-f84d94ed1cdf40fa9dd8958c706e5d69',
  '_parallel_build_trees-batch-3641e9275fa84e9b8e977c890f5d4660'),
 'tcp://127.0.0.1:44549': ('_parallel_build_trees-batch-1bc4abe0a9f44d648e9851f3628920d8',
  '_parallel_build_trees-batch-88cd3aa7c4324fa6aa9bb7cccf7b9baf'),
 'tcp://127.0.0.1:45417': ('_parallel_build_trees-batch-0f440e3dc3e449bc9fc80d31de714c1d',
  '_parallel_build_trees-batch-dfacedba845249f98ac1d3f2d6e42227')}
```
In this case, the batches are processed in parallel.

I suspect that this problem is somehow related to the choices make by the "AutoBalance" system in joblib. However, I have not been able to pin the exact problem. Is there some way to force joblib to use all dask workers, even for large problems?

**Comment by lesteve** [comment]
Maybe related to https://github.com/dask/dask/issues/5993?

**Comment by Chroxvi** [comment]
I have done some further testing and I am pretty sure the problem is to be found somewhere in `joblib` and its way of submitting tasks to dask. If I monkey patch to remove `joblib` from the fit method, everything works as expected - details below,

I have come up with a (very ugly) workaround that monkey patches the `RandomForestClassifier.fit` method here https://github.com/scikit-learn/scikit-learn/blob/b194674c4/sklearn/ensemble/_forest.py#L377 to avoid using `joblib` and instead directly submit the `_parallel_build_trees-batches` tasks to the dask scheduler (implemented as a contextmanager):
```python
import contextlib

import dask

@contextlib.contextmanager
def _dask_no_joblib_fit(model, X_delayed, y_delayed):
    import dask
    import dask.distributed

    def no_op_decoator(x):
        return x

    class DaskParallel:
        """
        Replacement of joblib.parallel.Parallel
        that directly computes tasks using Dask.
        """
        def __init__(self, *args, **kwargs):
            pass

        def __call__(self, tasks):
            with dask.distributed.worker_client() as client:
                task_futures = client.compute(tasks)
                result = client.gather(task_futures)

            return result

    org_parallel_build_trees = model.fit.__globals__['_parallel_build_trees']
    def wrap_parallel_build_trees(
            tree, forest, X, y, sample_weight, tree_idx, n_trees,
            verbose=0, class_weight=None, n_samples_bootstrap=None):
        """Replacement of _parallel_build_trees using delayed X, y."""
        result = dask.delayed(org_parallel_build_trees)(
            tree, forest, X_delayed, y_delayed, sample_weight, tree_idx,
            n_trees, verbose=verbose, class_weight=class_weight,
            n_samples_bootstrap=n_samples_bootstrap)
        return result

    org_delayed = model.fit.__globals__['delayed']
    org_Parallel = model.fit.__globals__['Parallel']
    model.fit.__globals__['delayed'] = no_op_decoator
    model.fit.__globals__['Parallel'] = DaskParallel
    model.fit.__globals__['_parallel_build_trees'] = wrap_parallel_build_trees
    yield
    model.fit.__globals__['delayed'] = org_delayed
    model.fit.__globals__['Parallel'] = org_Parallel
    model.fit.__globals__['_parallel_build_trees'] = org_parallel_build_trees
```
This allows me to fit the model using:
```python
X_delayed = dask.delayed(X)
y_delayed = dask.delayed(y)
with _dask_no_joblib_fit(model , X_delayed, y_delayed):
    model.fit(X, y)
```
which parallelizes nicely with on my dask cluster - even for large jobs.

**Comment by lesteve** [comment]
Thanks a lot for your input!

I haven't had time to look at this one yet, but I'd be interested to hear your feed-back on this recently opened PR: #1025.

**Comment by mrocklin** [comment]
I would be curious to know the size, in bytes, of this task: `dask_joblib_fit-23b412d6-a795-4b22-b425-5471684a0b51`

Dask decides where tasks should go based on the expected duration of that task, and the time it would take to move the dependencies of that task to a new machine.  If you have a large input, and fast tasks, then Dask may choose to put many tasks onto one machine, rather than spread that input around.  This would also explain why the balance tends to shift as you change the size of the problem.  

You might try explicitly scattering inputs with the `scatter=` keyword to `parallel_backend`.  See https://dask-ml.readthedocs.io/en/latest/joblib.html

**Comment by Chroxvi** [comment]
> I would be curious to know the size, in bytes, of this task: `dask_joblib_fit-23b412d6-a795-4b22-b425-5471684a0b51`

How would you obtain this size? Is there some dask functionality for this?

**Comment by Chroxvi** [comment]
> You might try explicitly scattering inputs with the `scatter=` keyword to `parallel_backend`. See https://dask-ml.readthedocs.io/en/latest/joblib.html

Using the `scatter=` keyword seems to partially solve the problem. For `n_estimators=20 >> n_workers=4`, I now get a distribution of the tasks across the workers that is more reasonable:
```python
{'tcp://127.0.0.1:34973': ('dask_joblib_fit-da67ac3b-11d2-4eee-a652-c2376d14f44a',
  '_parallel_build_trees-batch-b77327bb1ec146d5a7ed4a93732f5136',
  '_parallel_build_trees-batch-431a9cf844994d4aa58bdfca4a1aa1b5',
  '_parallel_build_trees-batch-8fb895a278bc49748b95951d79ae2f59',
  '_parallel_build_trees-batch-5d1c060ef37d43988ff3f770d2a33cfe'),
 'tcp://127.0.0.1:36631': ('_parallel_build_trees-batch-237d04b859304852adc812547725be7c',),
 'tcp://127.0.0.1:45283': ('_parallel_build_trees-batch-3cee20f84f2d49589b3a9b3c9519b475',),
 'tcp://127.0.0.1:46557': ('_parallel_build_trees-batch-d634f638955148de982d423ec7a4560a',
  '_parallel_build_trees-batch-9ca799c7ccf243d3a03f26c636f6a552')}
```
However, quite a few of the `_parallel_build_trees` tasks are scheduled on the worker that has the `dask_joblib_fit` task. Thus, for  `n_estimators=4 = n_workers=4`, you still see no parallelization:
```python
{'tcp://127.0.0.1:34473': ('dask_joblib_fit-88027b47-7e88-4f53-8f6c-184722fe080f',
  '_parallel_build_trees-batch-951727b1dc204cedab2c63152f71629b',
  '_parallel_build_trees-batch-2d857e95ed6147c29deeefc21a57d847',
  '_parallel_build_trees-batch-8da6e153d54f4d739010f23c4a917c5f',
  '_parallel_build_trees-batch-1c77bbe73d0147ea89a03819feab6c0b'),
 'tcp://127.0.0.1:35761': (),
 'tcp://127.0.0.1:37019': (),
 'tcp://127.0.0.1:42297': ()}
```
That being said, I would expect that the most common scenario is `n_estimators >> n_workers`, so using the pre-scattering is a big win.

**Comment by ogrisel** [comment]
@Chroxvi can you please tell us if the merge of the #1025 PR in joblib master helped with this issue?

Also tagging @pierreglaser here that has a bunch of work in progress branches to improve the dask/joblib backend, in particular related to auto-scattering and auto-batching.

This might still be a different issue (starving workers). Maybe we need to investigate if the default `pre_dispatch`  value used by the `Parallel` calls in scikit-learn is good or not.

**Comment by Chroxvi** [comment]
@ogrisel Pre-scattering is an excellent workaround for my use cases, so I forgot about #1025 - sorry about that. I will be happy to test my use cases against #1025. However, it will probably be a week or two before I can get at it.

**Comment by Chroxvi** [comment]
@ogrisel I have done some testing using
- dask version: 2.17.2
- joblib version: 0.15.1 (which includes #1025)
- scikit-learn version: 0.22.1

Results are somewhat mixed. My overall conlusions are:
- For small problem sizes everything still seems to work as expected.
- At some point, as the problem size increases, it seems that one worker in your dask cluster is, at least for a start, reserved for the calling task, e.g. `dask_joblib_fit` in this issue. That is, if your dask cluster has `N` workers, only `N-1` are used for computing the `_parallel_build_trees-batch` tasks created and submitted by joblib.
- For smaller problem sizes it does not seem to matter whether or not `scatter=[X, y]` is passed to `joblib.parallel_backend('dask')`.
- At some point, as the problem size increases, you still must pass `scatter=[X, y]` to `joblib.parallel_backend('dask')` in order to have the tasks submitted by joblib computed in parallel. If you dont use pre-scattering, only one or two tasks are computed at a time.

More details about my experiments are given below.

For the minimal example, I reported as part of this issue, for `n_samples=int(1e4)`, I get the following task scheduling (output of `client.processing()`):
```
{'tcp://127.0.0.1:36435': ('dask_joblib_fit-7bf75534-e30e-46f6-8d3e-ab6d8f7b6b52',
  '_parallel_build_trees-batch-487339ef54b441c4a8d3f8e8e7bb1bfe',
  '_parallel_build_trees-batch-fb75be633dae4748987ece9fb3ad6920'),
 'tcp://127.0.0.1:39619': ('_parallel_build_trees-batch-02ab1f39c2dc46b48f285b6b9a5f19b7',
  '_parallel_build_trees-batch-5b1f9d02c88b45b09ac00129aad81881'),
 'tcp://127.0.0.1:45411': ('_parallel_build_trees-batch-9bddb3bd26a84010ae1f35e8e38955fe',
  '_parallel_build_trees-batch-7f10c7f93ca64cc4acf6ed6905602f73'),
 'tcp://127.0.0.1:46567': ('_parallel_build_trees-batch-7b0eea313b1444d6bc70943a31530abb',
  '_parallel_build_trees-batch-0d9f0d0a5423422fb0fdf969b0d6885c')}
```
That is, everything (still) parallelizes nicely.
For `n_samples=int(1e6)`, I get:
```
{'tcp://127.0.0.1:34651': ('dask_joblib_fit-8f830100-d71f-4f59-983f-44df18361755',),
 'tcp://127.0.0.1:35907': ('_parallel_build_trees-batch-d6b184fc6e8f4617adc71b60189adcf1',
  '_parallel_build_trees-batch-ee80ae9ac3694cd9b0ae8600c1b14e40'),
 'tcp://127.0.0.1:38253': ('_parallel_build_trees-batch-54f1907c27f84342adabeef02393006c',
  '_parallel_build_trees-batch-4be1181fefda40c8a9e50c7a7ad13914'),
 'tcp://127.0.0.1:42231': ('_parallel_build_trees-batch-929de7e7cca44babb898f8297c7eef69',
  '_parallel_build_trees-batch-bf61a6cb12d04d538d730178c7297f18')}
``` 
That is, tasks are still processed in parallel, however, now only 3 out of 4 workers are used for computing the `_parallel_build_trees-batch-*` tasks. For this minimal example, it does not seem to matter whether or not `scatter=[X, y]` is passed to `joblib.parallel_backend('dask')`.

I also did a test on a real world use case. In this use case, `X.shape == (4_000_000, 157)` and I fit a `RandomForest` model with `n_esitmators=100` (resulting in 100  `_parallel_build_trees-batch-*` tasks being submitted by dask). I use `joblib.parallel_backend('dask')`, as in the minimal example with the dask distributed scheduler and 20 workers/processes (with one thread each) running on a compute server featuring 20 CPU cores / 40 HW threads.

If I don't pass  `scatter=[X, y]` to `joblib.parallel_backend('dask')`, only 1 to 3  `_parallel_build_trees-batch-*` tasks are running at a time on 1 to 2 workers. However, if I pass `scatter=[X, y]` to `joblib.parallel_backend('dask')`, 36 to 40  `_parallel_build_trees-batch-*` tasks are submitted and running on 19 workers at first. The last worker starts out idling, holding only the `dask_joblib_fit` task. Once, the first set of 19 tasks complete, the last worker seems to be assigned a   `_parallel_build_trees-batch-*` task, resulting in 20 tasks processing in parallel with about 40 tasks being submitted to the dask scheduler at a time.

**Comment by Chroxvi** [comment]
BTW, what happened to the documentation of the `scatter=[X, y]` kwarg to `joblib.parallel_backend('dask')`? I seem to remember it being part of https://dask-ml.readthedocs.io/en/latest/joblib.html. However it is no longer mentioned on that page.

**Comment by ogrisel** [comment]
I have tried the original reproducer script on a machine with 48 threads (24 physical cores) with `n_estimators=100` and I do not have any problem saturating the worker threads (48 running threads) with:

```
dask version: 2.20.0
joblib version: 0.16.0
scikit-learn version: 0.23.1
```

**Comment by ogrisel** [comment]
Actually I played with variants of that script and I could reproduce the bug several times (but it was random) then I changed the script to log more low level info and I cannot reproduce the problem any more.

@Chroxvi if you find a deterministic way to trigger the starving workers problem that would help me a lot.

**Comment by Chroxvi** [comment]
> Actually I played with variants of that script and I could reproduce the bug several times (but it was random) then I changed the script to log more low level info and I cannot reproduce the problem any more.

Yeah... Sounds a lot like the trouble I have had reproducing the issue in a deterministic way. I am about to go on vacation, but I will see what I can do when I get back around mid August.

**Comment by ogrisel** [comment]
Thanks @Chroxvi, no hurry. At least the problem seems less frequent that it used to be. However when trying to reproduce this I think I found maybe related issues when using joblib with large data arguments in nested parallel calls. I will open a dedicated issue as I am not 100% sure whether or not this is related to the original problem you reported.
